grove/deutsch_jozsa/deutsch_jozsa.py

```diff
@@ -0,0 +1,158 @@
+"""
+Module for the Deutsch-Jozsa Algorithm.
+Based off description in "Quantum Computation
+and Quantum Information" by Neilson and Chuang.
+"""
+
+import pyquil.quil as pq
+from pyquil.gates import *
+import numpy as np
+
+
+def oracle_function(unitary_funct, qubits, ancilla):
+    """
+    Defines an oracle that performs the following unitary transformation:
+    |x>|y> -> |x>|f(x) xor y>
+
+    :param matrix unitary_funct: Matrix representation of the function f, i.e. the
+                          unitary transformation that must be applied to a
+                          state |x> to put f(x) in qubit 0.
+    :param list qubits: List of qubits that enter as input |x>.
+    :param Qubit ancilla: Qubit to serve as input |y>.
+    :return: A program that performs the above unitary transformation.
+    :rtype: Program
+    """
+    if not is_unitary(unitary_funct):
+        raise ValueError, "Function must be unitary."
+    p = pq.Program()
+    scratch_bit = p.alloc()
+    bits_for_funct = [scratch_bit] + qubits
+    p.defgate("FUNCT", unitary_funct)
+    p.defgate("FUNCT-INV", unitary_funct.T.conj())
+    p.inst(tuple(['FUNCT'] + bits_for_funct))
+    p.inst(CNOT(qubits[0], ancilla))
+    p.inst(tuple(['FUNCT-INV'] + bits_for_funct))
+    return p
+
+
+def deutsch_jozsa(oracle, qubits, ancilla):
+    """
+    Implementation of the Deutsch-Jozsa Algorithm.
+
+    Can determine whether a function f mapping {0,1}^n to {0,1} is constant
+    or balanced, provided that it is one of them.
+
+    :param Program oracle: Program representing unitary application of function.
+    :param list qubits: List of qubits that enter as state |x>.
+    :param Qubit ancilla: Qubit to serve as input |y>.
+    :return: A program corresponding to the desired instance of the
+             Deutsch-Jozsa Algorithm.
+    :rtype: Program
+    """
+    p = pq.Program()
+
+    # Put ancilla bit into minus state
+    p.inst(X(ancilla), H(ancilla))
+
+    # Apply Hadamard, Unitary function, and Hadamard again
+    p.inst(map(H, qubits))
+    p += oracle
+    p.inst(map(H, qubits))
+    return p
+
+def unitary_function(mappings):
+    """
+    Creates a unitary transformation that maps each state to the values specified
+    in mappings.
+
+    Some (but not all) of these transformations involve a scratch qubit, so room for one is
+    always provided. That is, if given the mapping of n qubits, the calculated transformation
+    will be on n + 1 qubits, where the 0th is the scratch bit and the return value
+    of the function is left in the 1st.
```

Comment:
Since the scratch bit is always intialized to 0, the only input states to that matrix are |00> and |01>. Applying that matrix to either of those states will result in a state where the answer bit (qubit 1 in this case) is 0. 

In fact, that matrix is just a SWAP gate, which makes sense for the mapping [0, 0]. It swaps the scratch bit into the answer register so that the function always returns 0.